reflection/panic_handler.go

```diff
@@ -0,0 +1,84 @@
+package reflection
+
+import (
+	"reflect"
+	"sync"
+	"unsafe"
+	_ "unsafe"
+
+	"github.com/undefinedlabs/go-mpatch"
+)
+
+var (
+	patchOnPanic    *mpatch.Patch
+	mOnPanic        sync.Mutex
+	onPanicHandlers []func(e interface{})
+
+	patchOnExit   *mpatch.Patch
+	mOnExit       sync.Mutex
+	onExitHandler []func(e interface{})
+)
+
+// Adds a global panic handler (this handler will be executed before any recover call)
+func AddPanicHandler(fn func(interface{})) {
+	mOnPanic.Lock()
+	defer mOnPanic.Unlock()
+	if patchOnPanic == nil {
+		gp := lgopanic
+		np, err := mpatch.PatchMethodByReflect(reflect.Method{Func: reflect.ValueOf(gp)}, gopanic)
+		if err == nil {
+			patchOnPanic = np
+		}
+	}
+	onPanicHandlers = append(onPanicHandlers, fn)
+}
+
+// Adds a global panic handler before process kill (this handler will be executed if not recover is set before the process exits)
+func AddOnPanicExitHandler(fn func(interface{})) {
+	mOnExit.Lock()
+	defer mOnExit.Unlock()
+	if patchOnExit == nil {
+		gp := lpreprintpanics
+		np, err := mpatch.PatchMethodByReflect(reflect.Method{Func: reflect.ValueOf(gp)}, preprintpanics)
+		if err == nil {
+			patchOnExit = np
+		}
+	}
+	onExitHandler = append(onExitHandler, fn)
+}
+
+func gopanic(e interface{}) {
+	mOnPanic.Lock()
+	defer mOnPanic.Unlock()
+	for _, fn := range onPanicHandlers {
+		fn(e)
+	}
+	patchOnPanic.Unpatch()
+	defer patchOnPanic.Patch()
+	lgopanic(e)
+}
+
+func preprintpanics(p *_panic) {
+	mOnExit.Lock()
+	defer mOnExit.Unlock()
+	for _, fn := range onExitHandler {
+		fn(p.arg)
+	}
+	patchOnExit.Unpatch()
+	defer patchOnExit.Patch()
+	lpreprintpanics(p)
+}
+
+//go:linkname lgopanic runtime.gopanic
+func lgopanic(e interface{})
```

Comment:
Just to know, what we are making here?

Reply:
when you call `panic` in golang, internally the `runtime.gopanic` func is called, what we are doing here is is to get access to that private function with `lgopanic` using the linker so `lgopanic` and `runtime.gopanic` has the same address to the execution code of the func.

Reply:
we are going to need that function to patch it, by using an indirection with a var and the mpatch function.

Reply:
Golang Magic!